### ISSUE TITLE
[StableHLO] Broadcast op is generating incorrect output

### ISSUE THREAD

**Issue by mmanzoorTT** [tenstorrent/tt-mlir#846]
Broadcast op is generating incorrect output. A sample test case can be found in [tt-xla PR](https://github.com/tenstorrent/tt-xla/pull/15) where a `2x1xf32` tensor is broadcasted to `2x4xf32` tensor. 

```
generated output = Array([[1.5009999e-02, 6.7896283e-01, 0.0000000e+00, 0.0000000e+00],
       [8.4077908e-45, 0.0000000e+00, 1.4012985e-44, 0.0000000e+00]],      dtype=float32)
expected output = Array([[0.01501  , 0.01501  , 0.01501  , 0.01501  ],
       [0.6789628, 0.6789628, 0.6789628, 0.6789628]], dtype=float32)
```

**Comment by mmanzoorTT** [comment]
TTNN broadcasts a tensor implicitly to match the other tensor for operations and TTNN does not require explicit broadcast operation. So we take advantage of this implicit broadcasting for TTIR to TTNN conversion and remove TTIR broadcast operation. 

The failing test case performs a broadcast operation on input and returns the output; the broadcasted value is not used in any other operation. In this case, we cannot rely on implicit broadcast and require explicit broadcast operation.

**Comment by mmanzoorTT** [comment]
@sdjordjevicTT I am broadcasting 2x1 tensor to 2x4 tensor. Here is Stable HLO MLIR converted to TTIR MLIR converted to TTNN MLIR.
 
Stable HLO MLIR
```
module @jit_module_broadcast attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2x1xf32> {mhlo.sharding = "{replicated}"}) -> (tensor<2x4xf32> {jax.result_info = ""}) {
    %0 = stablehlo.broadcast_in_dim %arg0, dims = [0, 1] : (tensor<2x1xf32>) -> tensor<2x4xf32>
    return %0 : tensor<2x4xf32>
  }
}
```
TTIR MLIR
```
module @jit_module_broadcast attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2x1xf32> {mhlo.sharding = "{replicated}"}) -> (tensor<2x4xf32> {jax.result_info = ""}) {
    %0 = tensor.empty() : tensor<2x4xf32>
    %1 = "ttir.broadcast"(%arg0, %0) <{dimension = [0, 1], operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<2x1xf32>, tensor<2x4xf32>) -> tensor<2x4xf32>
    return %1 : tensor<2x4xf32>
  }
}
```
TTNN MLIR
```
module @jit_module_broadcast attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32, tt.device = #tt.device<workerGrid = #tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1)[s0, s1] -> (0, d0 floordiv s0, d1 floordiv s1, (d0 mod s0) * s1 + d1 mod s1), dramMap = (d0, d1)[s0, s1] -> (0, 0, ((((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 8192) mod 12, (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 98304 + (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) mod 8192), meshShape = , chipIds = [0]>, tt.system_desc = #tt.system_desc<[{arch = <wormhole_b0>, grid = 8x8, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32, l1_unreserved_base = 1024, erisc_l1_unreserved_base = 1024, dram_unreserved_base = 1024, dram_unreserved_end = 1073741824, physical_cores = {worker = [ 0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  1x0,  1x1,  1x2,  1x3,  1x4,  1x5,  1x6,  1x7,  2x0,  2x1,  2x2,  2x3,  2x4,  2x5,  2x6,  2x7,  3x0,  3x1,  3x2,  3x3,  3x4,  3x5,  3x6,  3x7,  4x0,  4x1,  4x2,  4x3,  4x4,  4x5,  4x6,  4x7,  5x0,  5x1,  5x2,  5x3,  5x4,  5x5,  5x6,  5x7,  6x0,  6x1,  6x2,  6x3,  6x4,  6x5,  6x6,  6x7,  7x0,  7x1,  7x2,  7x3,  7x4,  7x5,  7x6,  7x7] dram = [ 8x0,  9x0,  10x0,  8x1,  9x1,  10x1,  8x2,  9x2,  10x2,  8x3,  9x3,  10x3]}, supported_data_types = [<f32>, <f16>, <bf16>, <bfp_f8>, <bfp_bf8>, <bfp_f4>, <bfp_bf4>, <bfp_f2>, <bfp_bf2>, <u32>, <u16>, <u8>], supported_tile_sizes = [ 4x16,  16x16,  32x16,  4x32,  16x32,  32x32], num_cbs = 32}], [0], [3 : i32], [ 0x0x0x0]>} {
  func.func public @main(%arg0: tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x1xf32, #tt.memory_space<system>>>> {mhlo.sharding = "{replicated}"}) -> (tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>> {jax.result_info = ""}) {
    %0 = "ttnn.get_device"() <{mesh_shape = #ttnn<mesh_shape 1x1>}> : () -> !tt.device<<workerGrid = #tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1)[s0, s1] -> (0, d0 floordiv s0, d1 floordiv s1, (d0 mod s0) * s1 + d1 mod s1), dramMap = (d0, d1)[s0, s1] -> (0, 0, ((((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 8192) mod 12, (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 98304 + (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) mod 8192), meshShape = , chipIds = [0]>>
    %1 = "ttnn.to_layout"(%arg0) <{layout = #ttnn.layout<tile>}> : (tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x1xf32, #tt.memory_space<system>>>>) -> tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>
    %2 = "ttnn.to_device"(%1, %0) <{memory_config = #ttnn.memory_config<<interleaved>, <dram>, <<1x1>>>}> : (tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>, !tt.device<<workerGrid = #tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1)[s0, s1] -> (0, d0 floordiv s0, d1 floordiv s1, (d0 mod s0) * s1 + d1 mod s1), dramMap = (d0, d1)[s0, s1] -> (0, 0, ((((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 8192) mod 12, (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 98304 + (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) mod 8192), meshShape = , chipIds = [0]>>) -> tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>
    "ttnn.dealloc"(%1) : (tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>) -> ()
    %3 = "ttnn.from_device"(%2) : (tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>) -> tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>>
    "ttnn.dealloc"(%2) : (tensor<2x1xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #tt.memory_space<dram>>, interleaved>>) -> ()
    %4 = "ttnn.to_layout"(%3) <{layout = #ttnn.layout<row_major>}> : (tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>>) -> tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>>
    "ttnn.dealloc"(%3) : (tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>>) -> ()
    return %4 : tensor<2x4xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4xf32, #tt.memory_space<system>>>>
  }
}
```
